### PR TITLE
Rework Lavaland tunnel themes.

### PR DESCRIPTION
**Pull request by warriorstar-orion**
## What Does This PR Do
This PR:
- creates tunnel theme datums corresponding to the existing burrows, deeprock, and classic caves themes.
- uses the rust-g noise generation functions to create thematically appropriate tunnels.
- removes the "No Lava" mapping helper from lavaland.dmm proper, relying solely on area whitelisting to ensure lava placement for the base map.
- refactors some lava ruins, changing their shape and borders to make them fit better with the new procgen.

I tried to port the thematic elements of the cave themes as closely as possible.

## Why It's Good For The Game
For players: The new caves look a lot better and more natural than the old algorithm.

For coders: The existing cave generator was tightly integrated into a specific /turf subtype which contained all its behavior, which 1) was unwieldy and overcomplicated as the cave generation grew more complex and 2) was one of the turf types generated during _mineral_ generation, which would cause issues if trying to rework mineral generation in any way.

I believe using the rust functions will also be faster overall, though I didn't do any actual timing tests.

### Classic Caves

2024_08_24__21_35_52__Paradise Station 13

2024_08_24__21_35_58__Paradise Station 13

### Blocked Burrows

2024_08_24__21_40_46__Paradise Station 13

2024_08_24__21_40_52__Paradise Station 13

### Deadly Deeprock

2024_08_24__22_29_57__Paradise Station 13
2024_08_24__22_38_38__Paradise Station 13
2024_08_24__22_38_45__Paradise Station 13
2024_08_24__22_38_49__Paradise Station 13

## Testing
Generating lots and lots of maps.

Checking spawn rates for mobs/flora/megafauna in progress.

<hr>

### Declaration
- [X] I confirm that I either do not require [pre-approval](../CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.

<hr>

## Changelog
:cl:
tweak: Lavaland cave tunnels are now a lot more organic and natural looking.
/:cl:

### CODE REVIEW

File: code/game/turfs/simulated/floor/asteroid_floors.dm [hunk 1]
```diff
@@ -361,85 +198,6 @@ GLOBAL_LIST_INIT(megafauna_spawn_list, list(/mob/living/simple_animal/hostile/me
...
-			if(istype(loc, /area/lavaland/surface/outdoors/unexplored/danger)) //this is danger. it's boss time.
-				var/maybe_boss = pickweight(megafauna_spawn_list)
-				if(megafauna_spawn_list[maybe_boss])
-					randumb = maybe_boss
-			else //this is not danger, don't spawn a boss, spawn something else
-				randumb = pickweight(mob_spawn_list)
-		var/scanrange = 12
-		var/megafaunarange = 7
-		switch(SSmapping.cave_theme)
-			if(DEADLY_DEEPROCK)
```
Comment (Qwertytoforty) [inline code/game/turfs/simulated/floor/asteroid_floors.dm:398]: You'll probably want to see about adding this function back in for deadly deeprock, seems missing from the rework so the big rooms won't spawn as many mobs

**Comment by warriorstar-orion** [comment]
CI is failing because of that runtime in river generation (which happens before caves). Should hopefully be fixed with #26606.

File: code/controllers/subsystem/non_firing/SSmapping.dm [hunk 2]
```diff
@@ -53,8 +53,11 @@ SUBSYSTEM_DEF(mapping)
 	lavaland_theme = new lavaland_theme_type
 	log_startup_progress("We're in the mood for [initial(lavaland_theme.name)] today...") //We load this first. In the event some nerd ever makes a surface map, and we don't have it in lavaland in the event lavaland is disabled.
 
-	cave_theme = pick(BLOCKED_BURROWS, CLASSIC_CAVES, DEADLY_DEEPROCK)
-	log_startup_progress("We feel like [cave_theme] today...")
+	var/caves_theme_type = pick(subtypesof(/datum/caves_theme))
+	ASSERT(caves_theme_type)
+	caves_theme = new caves_theme_type
+	log_startup_progress("We feel like [initial(caves_theme.name)] today...")
```
Comment (AffectedArc07) [inline code/controllers/subsystem/non_firing/SSmapping.dm:59]: Do we need `initial` if we instance it on the line above?

Comment (warriorstar-orion) [reply code/controllers/subsystem/non_firing/SSmapping.dm]: Nope, let me fix that for both cave and planet themes.

File: code/modules/lavaland/caves_theme.dm [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+		return
+
+	for(var/turf/oasis_centroid in oasis_centroids)
+		if(get_dist(T, oasis_centroid) < oasis_padding)
+			return
+
+	oasis_centroids |= T
+	var/new_scan_range = rand(4, 7)
+	var/tempradius = rand(10, 15)
+	var/probmodifer = 43 * tempradius //Yes this is a magic number, it is a magic number that works well.
```
Comment (BiancaWilkson) [inline code/modules/lavaland/caves_theme.dm:151]: It's probably worth making this 43 into a define, for readability.

File: code/modules/lavaland/caves_theme.dm [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+			var/turf/changed = NT.ChangeTurf(/turf/simulated/floor/plating/asteroid/basalt/lava_land_surface)
+			if(prob(30))
+				lavaland_caves_spawn_mob(changed, new_scan_range, new_scan_range)
+			else if(prob(15))
+				lavaland_caves_spawn_flora(changed)
+
+	if(prob(50))
+		tempradius = round(tempradius / 3)
+		var/turf/oasis_lake = pickweight(lake_weights)
+		if(oasis_lake == /turf/simulated/floor/plating/asteroid)
```
Comment (lewcc) [inline code/modules/lavaland/caves_theme.dm:164]: Is this list holding types or values?

**Review by lewcc — COMMENTED** [review]
Looks pretty good to me, though I'd like more eyes on it from lavaland-familiar folks.

File: code/modules/lavaland/caves_theme.dm (hunk 3 above)
Comment (warriorstar-orion) [reply code/modules/lavaland/caves_theme.dm]: Done.

File: code/game/turfs/simulated/floor/asteroid_floors.dm (hunk 1 above)
Comment (warriorstar-orion) [reply code/game/turfs/simulated/floor/asteroid_floors.dm:412]: Done, I think, PTAL

File: code/modules/lavaland/caves_theme.dm (hunk 4 above)
Comment (warriorstar-orion) [reply code/modules/lavaland/caves_theme.dm]: It's holding types, I'll remove the /turf/ cast and change the name to make that clearer.

**Review by AffectedArc07 — APPROVED** [review]

**Review by DGamerL — APPROVED** [review]

**Pull request merged**